presto-spark-base/src/main/java/com/facebook/presto/spark/execution/LocalTask.java

```diff
@@ -0,0 +1,261 @@
+    private final AtomicReference<OutputBuffers> outputBuffers = new AtomicReference<>();
+    private final HttpClient httpClient;
+    private final TableWriteInfo tableWriteInfo;
+    private final RemoteTaskStats stats;
+    private final List<TaskSource> sources;
+    private TaskInfo taskInfo;
+    private final Executor executor;
+    private final ScheduledExecutorService updateScheduledExecutor;
+    @GuardedBy("this")
+    private ScheduledFuture<?> scheduledFuture;
+    private final JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec;
+    private final JsonCodec<PlanFragment> planFragmentCodec;
+    private final JsonCodec<TaskInfo> taskInfoCodec;
+
+    public LocalTask(
+            Session session,
+            TaskId taskId,
+            String nodeId,
+            URI location,
+            PlanFragment planFragment,
+            List<TaskSource> sources,
+            OutputBuffers outputBuffers,
+            HttpClient httpClient,
+            TableWriteInfo tableWriteInfo,
+            RemoteTaskStats stats,
+            Executor executor,
+            ScheduledExecutorService updateScheduledExecutor,
+            JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec,
+            JsonCodec<PlanFragment> planFragmentCodec,
+            JsonCodec<TaskInfo> taskInfoCodec)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(taskId, "taskId is null");
+        requireNonNull(nodeId, "nodeId is null");
+        requireNonNull(location, "location is null");
+        requireNonNull(planFragment, "planFragment is null");
+        requireNonNull(sources, "sources is null");
+        requireNonNull(outputBuffers, "outputBuffers is null");
+        requireNonNull(httpClient, "httpClient is null");
+
+        try (SetThreadName ignored = new SetThreadName("HttpRemoteTask-%s", taskId)) {
+            this.taskId = taskId;
+            this.session = session;
+            this.planFragment = planFragment;
+            this.outputBuffers.set(outputBuffers);
+            this.httpClient = httpClient;
+            this.tableWriteInfo = tableWriteInfo;
+            this.stats = stats;
+            this.sources = sources;
+            this.executor = executor;
+            this.updateScheduledExecutor = updateScheduledExecutor;
+            List<BufferInfo> bufferStates = outputBuffers.getBuffers()
+                    .keySet().stream()
+                    .map(outputId -> new BufferInfo(outputId, false, 0, 0, PageBufferInfo.empty()))
+                    .collect(toImmutableList());
+            this.taskInfo = createInitialTask(taskId, location, bufferStates, new TaskStats(DateTime.now(), null), nodeId);
+            this.taskUpdateRequestCodec = taskUpdateRequestCodec;
+            this.planFragmentCodec = planFragmentCodec;
+            this.taskInfoCodec = taskInfoCodec;
+        }
+    }
+
+    public TaskInfo getTaskInfo()
+    {
+        return taskInfo;
+    }
+
+    public void start()
```

Comment:
We can create a class called TaskInfoFetcher and delegate all http related stuff there. That way HttpRemoteTask will just rely on TaskInfoFetcher/TaskResultFetcher for low level polling and connection logic and simply deal with task related logic.